gui/src/WorkspaceView/controls/SurfaceSelectionControls.tsx

```diff
@@ -0,0 +1,164 @@
+            displayed: (selection.visibleSurfaceNames || []).includes(surface.name),
+            scalarFields: surface.scalarFields,
+            vectorFields: surface.vectorFields,
+            selectedScalarFieldName: selection.selectedSurfaceScalarFieldNames[surface.name],
+            selectedVectorFieldName: selection.selectedSurfaceVectorFieldNames[surface.name]
+        }
+    })), [surfacesWithFields, selection.selectedSurfaceVectorFieldNames, selection.selectedSurfaceScalarFieldNames, selection.visibleSurfaceNames])
+
+    return (
+        <div key="surfaces">
+            <h3>Surfaces</h3>
+            {
+                surfaces.map(surface => (
+                    <PerSurfaceControls
+                        name={surface.name}
+                        displayed={surface.displayed}
+                        scalarFields={surface.scalarFields}
+                        vectorFields={surface.vectorFields}
+                        selectedScalarFieldName={surface.selectedScalarFieldName}
+                        selectedVectorFieldName={surface.selectedVectorFieldName}
+                        visibilityCallback={handleToggleVisibleSurface}
+                        selectedScalarFieldCallback={handleToggleSelectedScalarField}
+                        selectedVectorFieldCallback={handleToggleSelectedVectorField}
+                    />
+                ))
+            }
+        </div>
+    )
+}
+
+const PerSurfaceControls: FunctionComponent<PerSurfaceControlsProps> = (surface: PerSurfaceControlsProps) => {
+    const { name, displayed, scalarFields, vectorFields, selectedScalarFieldName, selectedVectorFieldName, visibilityCallback, selectedVectorFieldCallback, selectedScalarFieldCallback } = surface
+    return (
+        <div key={name}>
+            <h3 key="surface">Surface: {name}</h3>
+            <div key="show">
+                <Checkbox
+                    value={name}
+                    checked={displayed}
+                    onChange={visibilityCallback}
+                /> show surface
+            </div>
+            {/* Vector field selector */}
+            {displayed && <FieldSelector
+                surfaceName={name}
+                fields={vectorFields}
+                selectedFieldName={selectedVectorFieldName}
+                type={'vector'}
+                callback={selectedVectorFieldCallback}
+            />}
+            {/* Scalar field selector */}
+            {displayed && <FieldSelector
+                surfaceName={name}
+                fields={scalarFields}
+                selectedFieldName={selectedScalarFieldName}
+                type={'scalar'}
+                callback={selectedScalarFieldCallback}
+            />}
+        </div>
+    )
+}
+
+const FieldSelector: FunctionComponent<FieldSelectorProps> = (props: FieldSelectorProps) => {
+    const { surfaceName, fields, selectedFieldName, type, callback } = props
+    if (fields.length === 0) return (<span />)
```

Comment:
There's probably a better way to do this--could move the conditional into the `PerSurfaceControls` component to begin with, or there may be a better way for this component to return a no-op itself without adding an empty tag?